File: demo/Delphi/Unit1.pas
unit Unit1;

interface

uses
  Windows, Classes, Controls, Forms,
  Highlight, Menus, StdCtrls, ComCtrls, ExtCtrls;

type
  TForm1 = class(TForm)
    PageControl1: TPageControl;
    TabSheet1: TTabSheet;
    TabSheet2: TTabSheet;
    TabSheet3: TTabSheet;
    ListBox1: TListBox;
    ListBox2: TListBox;
    ListBox3: TListBox;
    PopupMenu1: TPopupMenu;
    PopupMenu2: TPopupMenu;
    PopupMenu3: TPopupMenu;
    Redraw1: TMenuItem;
    Redraw2: TMenuItem;
    Redraw3: TMenuItem;
    N1: TMenuItem;
    N2: TMenuItem;
    N3: TMenuItem;
    Panel1: TPanel;
    Button1: TButton;
    procedure ListBox1DrawItem(Control: TWinControl; Index: Integer;
      Rect: TRect; State: TOwnerDrawState);
    procedure ListBox2DrawItem(Control: TWinControl; Index: Integer;
      Rect: TRect; State: TOwnerDrawState);
    procedure ListBox3DrawItem(Control: TWinControl; Index: Integer;
      Rect: TRect; State: TOwnerDrawState);
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure Redraw1Click(Sender: TObject);
    procedure Redraw2Click(Sender: TObject);
    procedure Redraw3Click(Sender: TObject);
    procedure Button1Click(Sender: TObject);
  private
    procedure OnClickM1(Sender: TObject);
    procedure OnClickM2(Sender: TObject);
    procedure OnClickM3(Sender: TObject);
    procedure LoadFromStream(ListBox: TListBox; Language: string);
  public
  end;

var
  Form1: TForm1;

implementation

{$R *.dfm}

var
  AsmLbId   : DWORD;
  HexLbId   : DWORD;
  DelphiLbId: DWORD;
  i         : integer;
  m1,m2,m3  : array of TMenuItem;

procedure TForm1.ListBox1DrawItem(Control: TWinControl; Index: Integer;
  Rect: TRect; State: TOwnerDrawState);
begin
  HighlightDrawItem(DelphiLbId, Index, Rect);
end;

procedure TForm1.ListBox2DrawItem(Control: TWinControl; Index: Integer;
  Rect: TRect; State: TOwnerDrawState);
begin
  HighlightDrawItem(AsmLbId, Index, Rect);
end;

procedure TForm1.ListBox3DrawItem(Control: TWinControl; Index: Integer;
  Rect: TRect; State: TOwnerDrawState);
begin
  HighlightDrawItem(HexLbId, Index, Rect);
end;

procedure TForm1.LoadFromStream(ListBox: TListBox; Language: string);
var
  RS: TResourceStream;
begin
  RS := TResourceStream.Create(hInstance, Language, 'EXAMPLE');
  try
    ListBox.Clear;
    ListBox.Items.LoadFromStream(RS);
  finally
    RS.Free;
  end;
end;

procedure TForm1.FormCreate(Sender: TObject);
var
  buf : array [0..4095] of AnsiChar;
begin
  if not InitHighlight then
  begin
    MessageBoxA(0, PAnsiChar(HighlightDll + ' not found'), 'Error', MB_ICONERROR);
    ExitProcess(0);
  end;
  ListBox1.Style := lbOwnerDrawFixed;
  ListBox2.Style := lbOwnerDrawFixed;
  ListBox3.Style := lbOwnerDrawFixed;
  DelphiLbId := CreateHighlight(ListBox1.Handle, ID_DELPHI);
  AsmLbId := CreateHighlight(ListBox2.Handle, ID_ASM);
  HexLbId := CreateHighlight(ListBox3.Handle, ID_HEX);
  LoadFromStream(ListBox1, 'DELPHI');
  LoadFromStream(ListBox2, 'ASM');
  LoadFromStream(ListBox3, 'HEX');
  //Для установки тем из меню
  for i := 0 to GetThemesCount(ID_DELPHI) - 1 do
  begin
    SetLength(m1, Length(m1) + 1);
    m1[i] := TMenuItem.Create(Self);
    GetThemeName(ID_DELPHI, DWORD(i), buf, Length(buf));
    m1[i].Caption := string(buf);
    m1[i].OnClick := OnClickM1;
    PopupMenu1.Items.Add(m1[i]);
  end;
  for i := 0 to GetThemesCount(ID_ASM) - 1 do
  begin
    SetLength(m2, Length(m2) + 1);
    m2[i] := TMenuItem.Create(Self);
    GetThemeName(ID_ASM, DWORD(i), buf, Length(buf));
    m2[i].Caption := string(buf);
    m2[i].OnClick := OnClickM2;
    PopupMenu2.Items.Add(m2[i]);
  end;
  for i := 0 to GetThemesCount(ID_HEX) - 1 do
  begin
    SetLength(m3, Length(m3) + 1);
    m3[i] := TMenuItem.Create(Self);
    GetThemeName(ID_HEX, DWORD(i), buf, Length(buf));
    m3[i].Caption := string(buf);
    m3[i].OnClick := OnClickM3;
    PopupMenu3.Items.Add(m3[i]);
  end;
end;

procedure TForm1.FormDestroy(Sender: TObject);
begin
  for i := Low(m1) to High(m1) do
    m1[i].Free;
  for i := Low(m2) to High(m2) do
    m2[i].Free;
  for i := Low(m3) to High(m3) do
    m3[i].Free;
  Finalize(m1);
  Finalize(m2);
  Finalize(m3);
  FreeHighlight;
end;

procedure TForm1.OnClickM1(Sender: TObject);
begin
  for i := Low(m1) to High(m1) do
    if (Sender as TMenuItem) = m1[i] then
    begin
      //ChangeGlobalTheme(ID_DELPHI, i);
      ChangeTheme(DelphiLbId, i);
    end;
end;

procedure TForm1.OnClickM2(Sender: TObject);
begin
  for i := Low(m2) to High(m2) do
    if (Sender as TMenuItem) = m2[i] then
    begin
      //ChangeGlobalTheme(ID_ASM, i);
      ChangeTheme(AsmLbId, i);
    end;
end;

procedure TForm1.OnClickM3(Sender: TObject);
begin
  for i := Low(m3) to High(m3) do
    if (Sender as TMenuItem) = m3[i] then
    begin
      //ChangeGlobalTheme(ID_HEX, i);
      ChangeTheme(HexLbId, i);
    end;
end;

procedure TForm1.Redraw1Click(Sender: TObject);
begin
  HighlightRedraw(DelphiLbId);
end;

procedure TForm1.Redraw2Click(Sender: TObject);
begin
  HighlightRedraw(AsmLbId);
end;

procedure TForm1.Redraw3Click(Sender: TObject);
begin
  HighlightRedraw(HexLbId);
end;

procedure TForm1.Button1Click(Sender: TObject);
var
  ModalResult : Integer;
begin
  ModalResult := SettingsShowModal(Handle);
end;

end.

File: demo/C++ Builder/Unit1.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "Unit1.h"
#include "Highlight.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
#pragma resource "data.res"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
	: TForm(Owner)
{
}
//---------------------------------------------------------------------------
DWORD AsmLbId;
DWORD HexLbId;
DWORD DelphiLbId;
int DelphiThemesCount;
int AsmThemesCount;
int HexThemesCount;
TMenuItem *m1[10];
TMenuItem *m2[10];
TMenuItem *m3[10];

void __fastcall TForm1::ListBox1DrawItem(TWinControl *Control, int Index, TRect &Rect,
          TOwnerDrawState State)
{
	HighlightDrawItem(DelphiLbId, Index, Rect, (rand() % 2 == 0 ? true : false));
}
//---------------------------------------------------------------------------

void __fastcall TForm1::ListBox2DrawItem(TWinControl *Control, int Index, TRect &Rect,
          TOwnerDrawState State)
{
	HighlightDrawItem(AsmLbId, Index, Rect, (rand() % 2 == 0 ? true : false));
}
//---------------------------------------------------------------------------

void __fastcall TForm1::ListBox3DrawItem(TWinControl *Control, int Index, TRect &Rect,
          TOwnerDrawState State)
{
	HighlightDrawItem(HexLbId, Index, Rect, (rand() % 2 == 0 ? true : false));
}
//---------------------------------------------------------------------------

void __fastcall TForm1::Redraw1Click(TObject *Sender)
{
	HighlightRedraw(DelphiLbId);
}
//---------------------------------------------------------------------------

void __fastcall TForm1::Redraw2Click(TObject *Sender)
{
	HighlightRedraw(AsmLbId);
}
//---------------------------------------------------------------------------

void __fastcall TForm1::Redraw3Click(TObject *Sender)
{
	HighlightRedraw(HexLbId);
}
//---------------------------------------------------------------------------

void __fastcall TForm1::LoadFromStream(TListBox *ListBox, char *Lang)
{
	TResourceStream *RS = new TResourceStream((int)HInstance, UnicodeString(Lang), L"EXAMPLE");
	ListBox->Items->Clear();
	ListBox->Items->LoadFromStream(RS);
	delete RS;
};

void __fastcall TForm1::FormCreate(TObject *Sender)
{

	if (!InitHighlight()){
      	MessageBoxA(NULL, "Highlight.dll not found", "Error", MB_ICONERROR);
		ExitProcess(NULL);
	};
	randomize;
	ListBox1->Style = lbOwnerDrawFixed;
	ListBox2->Style = lbOwnerDrawFixed;
	ListBox3->Style = lbOwnerDrawFixed;
	DelphiLbId = CreateHighlight((HWND)ListBox1->Handle, (DWORD)ID_DELPHI);
	AsmLbId    = CreateHighlight((HWND)ListBox2->Handle, (DWORD)ID_ASM);
	HexLbId    = CreateHighlight((HWND)ListBox3->Handle, (DWORD)ID_HEX);
	LoadFromStream(ListBox1, "DELPHI");
	LoadFromStream(ListBox2, "ASM");
	LoadFromStream(ListBox3, "HEX");
	DelphiThemesCount = GetThemesCount(ID_DELPHI);
	AsmThemesCount    = GetThemesCount(ID_ASM);
	HexThemesCount    = GetThemesCount(ID_HEX);
	LPCSTR buf[256];
	int i = 0;
	if (DelphiThemesCount > 0) {
		for (i = 0; i < DelphiThemesCount; i++){
			m1[i] = new TMenuItem(PopupMenu1->Items);
			GetThemeName(ID_DELPHI, (DWORD)i,(LPCSTR)buf, 256);
			m1[i]->Caption = UnicodeString((char *)buf);
			m1[i]->OnClick = OnClickM1;
			PopupMenu1->Items->Add(m1[i]);
		}
	}
	if (AsmThemesCount > 0) {
		for (i = 0; i < AsmThemesCount; i++){
			m2[i] = new TMenuItem(PopupMenu2->Items);
			GetThemeName(ID_ASM, (DWORD)i,(LPCSTR)buf, 256);
			m2[i]->Caption = UnicodeString((char *)buf);
			m2[i]->OnClick = OnClickM2;
			PopupMenu2->Items->Add(m2[i]);
		}
	}
	if (HexThemesCount > 0) {
		for (i = 0; i < HexThemesCount; i++){
			m3[i] = new TMenuItem(PopupMenu3->Items);
			GetThemeName(ID_HEX, (DWORD)i,(LPCSTR)buf, 256);
			m3[i]->Caption = UnicodeString((char *)buf);
			m3[i]->OnClick = OnClickM3;
			PopupMenu3->Items->Add(m3[i]);
		}
	}
}
//---------------------------------------------------------------------------

void __fastcall TForm1::FormDestroy(TObject *Sender)
{

	int i = 0;
	if (DelphiThemesCount > 0) {
		for (i = 0; i < DelphiThemesCount; i++){
			FreeAndNil(m1);
		}
	}
	if (AsmThemesCount > 0) {
		for (i = 0; i < AsmThemesCount; i++){
			FreeAndNil(m2);
		}
	}
	if (HexThemesCount > 0) {
		for (i = 0; i < HexThemesCount; i++){
			FreeAndNil(m3);
		}
	}
	FreeHighlight();
}
//---------------------------------------------------------------------------

void __fastcall TForm1::OnClickM1(TObject *Sender)
{
	if (DelphiThemesCount > 0) {
		for (int i = 0; i < DelphiThemesCount; i++){
			if ((Sender) == m1[i]){
				//ChangeGlobalTheme(ID_DELPHI, i);
				ChangeTheme(DelphiLbId, i);
			}
		}
	}
}

void __fastcall TForm1::OnClickM2(TObject *Sender)
{
	if (AsmThemesCount > 0) {
		for (int i = 0; i < AsmThemesCount; i++){
			if ((Sender) == m2[i]){
				//ChangeGlobalTheme(ID_ASM, i);
				ChangeTheme(AsmLbId, i);
			}
		}
	}
}

void __fastcall TForm1::OnClickM3(TObject *Sender)
{
	if (HexThemesCount > 0) {
		for (int i = 0; i < HexThemesCount; i++){
			if ((Sender) == m3[i]){
				//ChangeGlobalTheme(ID_HEX, i);
				ChangeTheme(HexLbId, i);
			}
		}
	}
}

void __fastcall TForm1::Button1Click(TObject *Sender)
{
	int ModalResult = SettingsShowModal(Handle);
}
//---------------------------------------------------------------------------

